### PR TITLE
COS (Tencent Cloud Object Storage) integration

### PR DESCRIPTION
**Pull request by fastio**
I hereby agree to the terms of the CLA available at: https://yandex.ru/legal/cla/?lang=en

Changelog category (leave one):
- New Feature

To allow import and export to COS compatible storage.

### CODE REVIEW

File: src/TableFunctions/TableFunctionCOS.h [hunk 1]
```diff
@@ -0,0 +1,47 @@
...
+{
+
+class Context;
+
+/* cos(source, access_key_id, secret_access_key, format, structure) - creates a temporary storage for a file in COS 
+ */
+class TableFunctionCOS : public ITableFunction
+{
+public:
+    static constexpr auto name = "cos";
```
Comment (alexey-milovidov) [inline src/TableFunctions/TableFunctionCOS.h:20]: `cos` is quite ambiguous name for function.
People may think it is somehow related to "cosine".

Maybe we should name it `tencentCOS`?

PS. https://intl.cloud.tencent.com/product/cos
> fully compatible with S3's API

How is it different actually? Why we cannot just use table function `s3`?

Comment (fastio) [reply src/TableFunctions/TableFunctionCOS.h:20]: > Maybe we should name it tencentCOS?

We should rename it to cosn which is another name of COS.

> PS. https://intl.cloud.tencent.com/product/cos
> 
> fully compatible with S3's API
> 
> How is it different actually? Why we cannot just use table function s3?

Yes, COS is compatible with S3's API.
But the cos has own sytle of endopoint. 
The current implementation of S3::URI can not parse the url of COS correctly.
I reuse the most of the codes of implementation of S3 integration, and write 
the codes to parse the COS url in COSCommon.*

The users of Tencent Cloud are very familiar with COS. We have integrate the 
COS to appache/Hive, HDFS, Presto etc. It's better  to be consistent here, 
use the the symbol COS(or COSN)

Comment (alexey-milovidov) [reply src/TableFunctions/TableFunctionCOS.h:20]: Ok, let's name it cosn.

Is there any difference in addition to URI style?
Maybe we can reduce code duplication by reusing most of the code from s3 implementation with only additions in
- new function name;
- some flags for URI parsing.

Comment (fastio) [reply src/TableFunctions/TableFunctionCOS.h:20]: Thanks.

> Is there any difference in addition to URI style?

E.g. https://hk-123456789.cos.ap-hongkong.myqcloud.com/test.dat

* COS only supports http/https scheme.
* hk-123456789: This is bucket name of COS, it should continue digital string which is APPID of Tencent cloud user.
* ap-hongkong: This is region name of COS.
* cos.ap-hongkong.myqcloud.com: This is the endpoint of COS service in Hong Kong region. 

> Maybe we can reduce code duplication by reusing most of the code from s3 implementation with only additions in
> 
> new function name;
> some flags for URI parsing.

Strongly agree.

**Comment by alexey-milovidov** [comment]
Superseded in #12386

**Pull request closed without merging**